---
params:
  title: "bulk RNAseq QC"
output: 
  rmdformats::robobook:
    fig_width: 9
    toc_depth: 3
    css: "stylesheets/robobook.css"
title: "`r params$title`"
author: "EPRC Gencore"
date:  "`r Sys.Date()`"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(warning = FALSE, message = FALSE, echo=FALSE, fig.width = 6, fig.height = 4, dev='png')
```

# Study overview

All metadata gleaned from sample names. Timepoints are calculated by the order of
sample dates for each individual/tissue combo. So the first timepoint will be the
earliest sample for that individual/tissue. However, there appears to be an error
in one samples labeling: Z15228 has samples from only 1 timepoint, and Z15229 has
samples from 2 timepoints. 

**Are some of these samples mislabeled? Should Z15228 and Z15229 be the same individual?**

If this is the case, the timepoint calculation will be resolved as there will 
be 3 samples for that individual.

```{r}
sample_metadata_table
```

# QC

## Mapping

Outcomes from mapping reads to reference genome. Reads need to be 'identified' or successfully mapped to a single feature to be used for gene expression analysis.

:::: {class='fluid-row'}

::: {class='col-md-6'}

### Mmul10

```{r, fig.width=8, fig.height=4}
mmul10_analysis$plots$mapping_bins_plot
```

:::

::: {class='col-md-6'}

### Mnem

```{r, fig.width=8, fig.height=4}
mnem_analysis$plots$mapping_bins_plot
```

:::
::::

## Relative Log Expression (RLE) normalization

Library size normalization of read counts with hypothesis that most genes are not differentially expressed

:::: {class='fluid-row'}

::: {class='col-md-6'}

### Mmul10

```{r, fig.width = 12, fig.height=8}
mmul10_analysis$plots$RLE_raw 
mmul10_analysis$plots$RLE_norm 
```

:::

::: {class='col-md-6'}

### Mnem

```{r, fig.width = 12, fig.height=8}
mnem_analysis$plots$RLE_raw 
mnem_analysis$plots$RLE_norm 
```

:::
::::

## PCA

Exploring projections of expression profiles onto low dimensional space using a limited number of highly variable features. This will give us an estimate of similarity between samples, but will not account for all genes.

:::: {class='fluid-row'}

::: {class='col-md-6'}

### Mmul10

#### PBMC

```{r, fig.width=6, fig.height=5}
mmul10_analysis$plots$pbmc_pca
```

#### Rectal biopsy

```{r, fig.width=6, fig.height=5}
mmul10_analysis$plots$rb_pca
```

:::

::: {class='col-md-6'}

### Mnem

#### PBMC

```{r, fig.width=6, fig.height=5}
mnem_analysis$plots$pbmc_pca
```

#### Rectal biopsy

```{r, fig.width=6, fig.height=5}
mnem_analysis$plots$rb_pca
```

:::

::::
