---
title: "p23259 Gerardo Analysis"
author: "ENPRC Genomics Core"
date:  "`r Sys.Date()`"
output: 
  rmdformats::robobook:
    fig_width: 9
    css: "stylesheets/robobook.css"
---

```{r setup, include=FALSE}
library(knitr)
knitr::opts_chunk$set(warning = FALSE,
                      message = FALSE, 
                      echo=FALSE, 
                      fig.width = 8, 
                      fig.height = 4)
```

# Differentially expressed genes 

This section of the report analyses gene expression between groups of samples within the study. Changes in gene expression between groups are usually reported on a log2 scale. The significance of a detected change in expression is reported in a raw/nominal p-value and a multiple-testing corrected/adjusted p-value that accounts for the large number of tests conducted in differential gene expression (DGE) and the inherent false positive rate assumed with frequentest p-values.

Each tab has results from a specific comparison within the study. Each tab shows 

1. A summary of significant DGE

2. A volcano plot of DGE

3. A table of all significant DGE

The order of terms in tab labeling informs how to interpret results. For log-fold change and enrichment, the first listed group is the numerator and the second is the denominator. So in a comparison of X vs Y, a positive log-fold change means the gene is more highly expressed in X, while a negative change means the gene is more highly expressed in Y. 

## PBMC {.tabset}

```{r, results='asis'}
out <- NULL
model_results <- model_results.pbmc
obj <- obj.pbmc
for (i in names(model_results.pbmc)) {
  ## Edit this file to change how tabs are generated for each result
  out = c(out, knit_expand(file = here('R/results_chunk_template.Rmd')))
}

cat(knit_child(text = out, quiet = TRUE), sep = '\n')
```

## Rectal biopsy {.tabset}

```{r, results='asis'}
out <- NULL
model_results <- model_results.rb
obj <- obj.rb
for (i in names(model_results.rb)) {
  ## Edit this file to change how tabs are generated for each result
  out = c(out, knit_expand(file = here('R/results_chunk_template.Rmd')))
}

cat(knit_child(text = out, quiet = TRUE), sep = '\n')
```

***

# Genes of interest {.tabset}

These heatmaps show the relative expression (regularized and log2 transformed) of genes for each sample normalized to the pre-treatment baseline for that sample. If you have genes you are interested in seeing from your study, we can plot them here.

## Top genes from whole study {.tabset}

### PBMC

```{r, fig.height=4}
geneList <- unique(unlist(lapply(model_results.pbmc, getTopNGenes, N=10, direction='equal')))

data_to_plot <- assays(obj.pbmc)$rld
data_to_plot <- normalizeCountsForHeatmapByIndividual(
  assay(data_to_plot),
  data_to_plot@colData,
  group_var = 'Timepoint', baseline = '1', 
  individual_var = 'Individual',
  remove_baseline = TRUE)
heatmapFromGenelist(geneList = geneList, 
                    data = data_to_plot,
                    column_split = c(rep(1,5), rep(2,5)),
                    slice_labels = c('W12', 'W18'),
                    slice_labels_rot = 0, 
                    column_labels = gsub('.+(Z[0-9]+).+','\\1', colnames(data_to_plot)))
```

***

### RB

```{r, fig.height=4}
geneList <- unique(unlist(lapply(model_results.rb, getTopNGenes, N=10, direction='equal')))

data_to_plot <- assays(obj.rb)$rld
data_to_plot <- normalizeCountsForHeatmapByIndividual(
  assay(data_to_plot),
  data_to_plot@colData,
  group_var = 'Timepoint', baseline = '1', 
  individual_var = 'Individual',
  remove_baseline = TRUE)
heatmapFromGenelist(geneList = geneList, 
                    data = data_to_plot,
                    column_split = c(rep(1,5), rep(2,5)),
                    slice_labels = c('W12', 'W18'),
                    slice_labels_rot = 0, 
                    column_labels = gsub('.+(Z[0-9]+).+','\\1', colnames(data_to_plot)))
```

***

# GSEA {.tabset}

The order of terms in tab labeling informs how to interpret results. For log-fold change and enrichment, the first listed group is the numerator and the second is the denominator. So in a comparison of X vs Y, a positive enrichment relates to enrichment in X, and negative enrichment indicates greater enrichment in Y. 

Gene Set Enrichment Analysis (GSEA) is a statistical approach to test systematic changes in groups of genes corresponding to biological pathways. A pathway is generally 'enriched' in one study group if the genes in that pathway are systematically up or down regulated in a manner and magnitude that is unlikely to be due to chance. Read more about GSEA [here](https://www.gsea-msigdb.org/gsea/index.jsp).

GSEA results are separated by the tissue type. Within each tissue, there is a plot
of the top pathways for each comparison and a table with more extensive results, limited
to the top 25 from each major source (Hallmark, Reactome, Kegg, GO). There is 
also an example of a joint GSEA plot which shows multiple pathways over all comparisons.
A few example pathways are selected. These are meant to be exploratory/examples. 
We can re-generate these plots for specific pathways you want to see. 

## PBMC

### Individual {.tabset}

```{r}
chunktext <- "
#### {{i}}
\n```{r, fig.height=10}
gseaDotplot_single(gsea_results_pbmc[['{{i}}']], signif_only = FALSE)\n```
\n```{r}
generateGSEADatatable(gsea_results_pbmc[['{{i}}']])\n```
"
```

```{r, results='asis'}
out <- NULL
for (i in names(model_results.pbmc)) {
  out = c(out, knit_expand(text = chunktext))
}

cat(knit_child(text = out, quiet = TRUE), sep = '\n')
```

### Joint

```{r}
gseaDotplot_joint(joint_GSEA_results_pbmc)
```

## RB

### Individual {.tabset}

```{r}
chunktext <- "
#### {{i}}
\n```{r, fig.height=10}
gseaDotplot_single(gsea_results_rb[['{{i}}']], signif_only = FALSE)\n```
\n```{r}
generateGSEADatatable(gsea_results_rb[['{{i}}']])\n```
"
```

```{r, results='asis'}
out <- NULL
for (i in names(model_results.rb)) {
  out = c(out, knit_expand(text = chunktext))
}

cat(knit_child(text = out, quiet = TRUE), sep = '\n')
```

### Joint

```{r}
gseaDotplot_joint(joint_GSEA_results_rb)
```
